perl-package/AI-MXNet/lib/AI/MXNet/Metric.pm

```diff
@@ -79,7 +79,7 @@ use overload '""' => sub {
 has 'name'       => (is => 'rw', isa => 'Str');
 has 'num'        => (is => 'rw', isa => 'Int');
 has 'num_inst'   => (is => 'rw', isa => 'Maybe[Int|ArrayRef[Int]]');
-has 'sum_metric' => (is => 'rw', isa => 'Maybe[Num|ArrayRef[Num]]');
+has 'sum_metric' => (is => 'rw', isa => 'Maybe[Num|ArrayRef[Num]|PDL]');
```

Comment:
Why is this needed ? Is $pdl->sum (and other similar methods) produce pdl object now ? This change is a bit radical and will require additional tests to see of something that is not covered by the CI is not broken.

Reply:
Yes, that is correct, they return a PDL now. This is to make sure that types such as complex numbers and BAD values can be correctly returned. To get a single Perl scalar out of a PDL, you can do `$pdl->sclr`. Though for many situations, this should work like a Perl scalar due to operator overloading.

Reply:
What type of tests would you like to see?

Reply:
May I ask you locally run scripts from AI-MXNet/examples/ and see if they run and produce sane output ?

Reply:
== examples/calculator.pl ==

```
fullyconnected4_bias -> [0.675975]
fullyconnected4_weight -> 
[
 [ 1.00116 0.999985]
]

fullyconnected5_bias -> [4.85674e-05]
fullyconnected5_weight -> 
[
 [0.000844207 0.000230625    0.508509]
]

fullyconnected6_bias -> [0.385759]
fullyconnected6_weight -> 
[
 [      1      -1]
]

fullyconnected7_bias -> [-2.62715e-06]
fullyconnected7_weight -> 
[
 [-1.31198e-05  8.74287e-06     0.679938]
]

12345 + 54321 ≈ 66742.1796875
188 - 88 ≈ 99.8591766357422
250 * 2 ≈ 503.255798339844
250 / 2 ≈ 124.998970031738
```

== examples/mnist.pl ==

```
Epoch[4] Batch [200]	Speed: 59678.68 samples/sec	Train-accuracy=0.974726
Epoch[4] Batch [400]	Speed: 57716.06 samples/sec	Train-accuracy=0.973350
Epoch[4] Train-accuracy=0.974233
Epoch[4] Time cost=1.057
Epoch[4] Validation-accuracy=0.967800
Epoch[5] Batch [200]	Speed: 52405.81 samples/sec	Train-accuracy=0.977214
Epoch[5] Batch [400]	Speed: 57804.15 samples/sec	Train-accuracy=0.976100
Epoch[5] Train-accuracy=0.976950
Epoch[5] Time cost=1.079
Epoch[5] Validation-accuracy=0.966500
Epoch[6] Batch [200]	Speed: 57043.78 samples/sec	Train-accuracy=0.977015
Epoch[6] Batch [400]	Speed: 57357.53 samples/sec	Train-accuracy=0.975800
Epoch[6] Train-accuracy=0.976433
Epoch[6] Time cost=1.052
Epoch[6] Validation-accuracy=0.963000
Epoch[7] Batch [200]	Speed: 45378.31 samples/sec	Train-accuracy=0.977711
Epoch[7] Batch [400]	Speed: 55753.34 samples/sec	Train-accuracy=0.976350
Epoch[7] Train-accuracy=0.977467
Epoch[7] Time cost=1.144
Epoch[7] Validation-accuracy=0.963700
Epoch[8] Batch [200]	Speed: 54546.75 samples/sec	Train-accuracy=0.980448
Epoch[8] Batch [400]	Speed: 55774.47 samples/sec	Train-accuracy=0.980000
Epoch[8] Train-accuracy=0.980217
Epoch[8] Time cost=1.090
Epoch[8] Validation-accuracy=0.962900
Epoch[9] Batch [200]	Speed: 48810.85 samples/sec	Train-accuracy=0.980050
Epoch[9] Batch [400]	Speed: 54945.51 samples/sec	Train-accuracy=0.982550
Epoch[9] Train-accuracy=0.981850
Epoch[9] Time cost=1.115
Epoch[9] Validation-accuracy=0.969500
```

== examples/sparse/matrix_factorization/train.pl ==

```
Epoch[2] Batch [60300]	Speed: 50291.13 samples/sec	Train-mse=0.706833
Epoch[2] Batch [60400]	Speed: 54047.43 samples/sec	Train-mse=0.723078
Epoch[2] Batch [60500]	Speed: 53646.70 samples/sec	Train-mse=0.741219
Epoch[2] Batch [60600]	Speed: 52929.93 samples/sec	Train-mse=0.706861
Epoch[2] Batch [60700]	Speed: 53583.14 samples/sec	Train-mse=0.734509
Epoch[2] Batch [60800]	Speed: 50907.01 samples/sec	Train-mse=0.735880
Epoch[2] Batch [60900]	Speed: 39899.38 samples/sec	Train-mse=0.758780
Epoch[2] Batch [61000]	Speed: 48115.77 samples/sec	Train-mse=0.693772
Epoch[2] Batch [61100]	Speed: 53833.52 samples/sec	Train-mse=0.714911
Epoch[2] Batch [61200]	Speed: 53998.88 samples/sec	Train-mse=0.731971
Epoch[2] Batch [61300]	Speed: 51153.75 samples/sec	Train-mse=0.732873
Epoch[2] Batch [61400]	Speed: 51454.19 samples/sec	Train-mse=0.715944
Epoch[2] Batch [61500]	Speed: 51353.83 samples/sec	Train-mse=0.715877
Epoch[2] Batch [61600]	Speed: 53121.93 samples/sec	Train-mse=0.703896
Epoch[2] Batch [61700]	Speed: 53345.31 samples/sec	Train-mse=0.700082
Epoch[2] Batch [61800]	Speed: 48247.09 samples/sec	Train-mse=0.734545
Epoch[2] Batch [61900]	Speed: 42918.29 samples/sec	Train-mse=0.714020
Epoch[2] Batch [62000]	Speed: 49330.02 samples/sec	Train-mse=0.722819
Epoch[2] Batch [62100]	Speed: 53999.32 samples/sec	Train-mse=0.736258
Epoch[2] Batch [62200]	Speed: 51953.96 samples/sec	Train-mse=0.737786
Epoch[2] Batch [62300]	Speed: 51416.39 samples/sec	Train-mse=0.698692
Epoch[2] Batch [62400]	Speed: 53206.32 samples/sec	Train-mse=0.716579
Epoch[2] Batch [62500]	Speed: 52938.28 samples/sec	Train-mse=0.733243
Preparing data iterators for ./data/ml-10M100K/r1.train ... 
Preparing data iterators for ./data/ml-10M100K/r1.test ... 
Training started ...
epoch 0, eval MSE = 0.99958598613739 
epoch 1, eval MSE = 1.03603255748749 
epoch 2, eval MSE = 1.07017529010773 
Training completed.
```

Reply:
Looks good